inmoose/utils/splines.py

```diff
@@ -0,0 +1,196 @@
+            :code:`False`.
+        boundary_knots : array_like, optional
+            boundary points at which to impose the natural boundary conditions and
+            anchor the B-spline basis (default the range of the data). If both
+            :code:`knots` and :code:`boundary_knots` are supplied, the basis
+            parameters do not depend on :code:`x`. Data can extend beyond
+            :code:`boundary_knots`.
+
+        Returns
+        -------
+        ndarray
+            a matrix of dimension :code:`(len(x), df)`, where :code:`df =
+            len(knots)-1-intercept` if :code:`df` was not supplied.
+        """
+        self.include_intercept = include_intercept
+        x = np.asarray(x)
+        if boundary_knots is None:
+            boundary_knots = [np.min(x), np.max(x)]
+            outside = False
+        else:
+            boundary_knots = list(np.sort(boundary_knots))
+            out_left = x < boundary_knots[0]
+            out_right = x > boundary_knots[1]
+            outside = out_left | out_right
+        self.boundary_knots = boundary_knots
+
+        if df is not None and knots is None:
+            nIknots = df - 1 - include_intercept
+            if nIknots < 0:
+                nIknots = 0
+                logging.warnings.warn("df was too small, used {1+include_intercept}")
+
+            if nIknots > 0:
+                knots = np.linspace(0, 1, num=nIknots + 2)[1:-1]
+                knots = np.quantile(x, knots)
+        else:
+            nIknots = len(knots)
+        self.knots = knots
+
+        Aknots = np.sort(np.concatenate((boundary_knots * 4, knots)))
+
+        if np.any(outside):
+            basis = np.empty((x.shape[0], nIknots + 4), dtype=float)
+            if np.any(out_left):
+                k_pivot = boundary_knots[0]
+                xl = np.ones((np.sum(out_left), 2))
+                xl[:, 1] = x[out_left] - k_pivot
+                tt = spline_design(Aknots, [k_pivot, k_pivot], 4, [0, 1])
+                basis[out_left, :] = xl @ tt
+            if np.any(out_right):
+                k_pivot = boundary_knots[1]
+                xr = np.ones((np.sum(out_right), 2))
+                xr[:, 1] = x[out_right] - k_pivot
+                tt = spline_design(Aknots, [k_pivot, k_pivot], 4, [0, 1])
+                basis[out_right, :] = xr @ tt
+            inside = ~outside
+            if np.any(inside):
+                basis[inside, :] = spline_design(Aknots, x[inside], 4)
+        else:
+            basis = spline_design(Aknots, x, 4)
+
+        const = spline_design(Aknots, boundary_knots, 4, [2, 2])
+
+        if include_intercept is False:
+            basis = basis[:, 1:]
+            const = const[:, 1:]
+
+        qr_const = np.linalg.qr(const.T, mode="complete")[0]
+        self.basis = (qr_const.T @ basis.T).T[:, 2:]
+
+    def predict(self, newx):
+        """
+        Evaluate the spline basis at given values
```

Comment:
Missing arg and return. Do you add return type usually?

Reply:
documentation updated